src/main/java/org/mskcc/cmo/messaging/impl/JSGatewayImpl.java

```diff
@@ -297,4 +296,43 @@ public PublishingQueueTask(String msgId, String subject, Object payload) {
             this.payload = payload;
         }
     }
+
+    @Override
+    public String request(String subject, Object message) {
+        if (natsConnection == null) {
+            LOG.error("Error initializing NATS connection");
+        }
+        try {
+            String msg = mapper.writeValueAsString(message);
+            Message reply = natsConnection.request(subject, msg.getBytes(), Duration.ofSeconds(10));
+            if (reply == null) {
+                LOG.error("No reply received for a request using NATS connection");
+            } else {
+                return reply.getData().toString();
+            }
+        } catch (Exception ex) {
+            LOG.error("Error during attempt to send a request using NATS connection", ex);
+        }
+        return null;
+    }
+
+    @Override
+    public void reply(String subject, Object message) {
+        if (natsConnection == null) {
+            LOG.error("Error initializing NATS connection");
+        }
+        try {
+            Subscription sub = natsConnection.subscribe(subject);
```

Comment:
Should this also take a custom id like we had to set up for our NATS publisher?

Reply:
To catch the duplicate messages?